File: 2025/SoyoungJung/week10/2025-08-10-House-Robber.ts
function rob(nums: number[]): number {
  if (nums.length === 0) return 0;
  if (nums.length === 1) return nums[0];

  let prev2 = nums[0];
  let prev1 = Math.max(nums[0], nums[1]);

  for (let i = 2; i < nums.length; i++) {
    const temp = prev1;
    prev1 = Math.max(prev1, prev2 + nums[i]);
    prev2 = temp;
  }
  return prev1;
}

File: 2025/HyeGyung/2025-08-10-House-Robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 0:
            return 0
        if n == 1:
            return nums[0]

        prev2 = nums[0]  # dp[i-2]
        prev1 = max(nums[0], nums[1])  # dp[i-1]

        for i in range(2, n):
            cur = max(prev1, prev2 + nums[i])  # dp[i]
            prev2, prev1 = prev1, cur

        return prev1
